client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentEventReaderImpl.java

```diff
@@ -0,0 +1,380 @@
+
+        this.readState = new ReadState();
+    }
+
+    @Override
+    public Segment getSegmentId() {
+        return asyncInput.getSegmentId();
+    }
+
+    @Override
+    @Synchronized
+    public void close() {
+        log.trace("Closing {}", this);
+        readState.cancel();
+        asyncInput.close();
+    }
+
+    @Override
+    public boolean isClosed() {
+        return asyncInput.isClosed();
+    }
+
+    // region Cursor
+
+    @Override
+    @Synchronized
+    public void setOffset(long offset) {
+        log.trace("SetOffset {}", offset);
+        Preconditions.checkArgument(offset >= 0);
+        Exceptions.checkNotClosed(asyncInput.isClosed(), this);
+        if (offset != this.cursorOffset) {
+            readState.cancel();
+            this.cursorOffset = offset;
+        }
+    }
+
+    @Override
+    @Synchronized
+    public long getOffset() {
+        return cursorOffset;
+    }
+
+    @Override
+    @Synchronized
+    public void setEndOffset(long offset) {
+        log.trace("SetEndOffset {}", offset);
+        Preconditions.checkArgument(offset >= 0);
+        Exceptions.checkNotClosed(asyncInput.isClosed(), this);
+        this.endOffset = offset;
+    }
+
+    @Override
+    @Synchronized
+    public long getEndOffset() {
+        return endOffset;
+    }
+
+    @Override
+    @Synchronized
+    public CompletableFuture<ByteBuffer> readAsync() throws EndOfSegmentException {
+
+        // prepare a promise
+        if (cursorOffset >= this.endOffset) {
+            log.debug("All events up to the configured end offset:{} have been read", endOffset);
+            throw new EndOfSegmentException(END_OFFSET_REACHED);
+        }
+        CompletableFuture<ByteBuffer> promise = readState.reset(cursorOffset);
+        promise.whenComplete(this::readCompleted);
+
+        // kick off or continue processing the request
+        issueRequestIfNeeded();
+
+        return promise;
+    }
+
+    @Synchronized
+    private void readCompleted(ByteBuffer result, Throwable th) {
+        if (th != null) {
+            // an event could not be read; don't update the cursor position.
+            log.trace("Failed to read an event @ " + cursorOffset, th);
+
+            // cancel any outstanding request now to ensure that handleRequest doesn't affect the read state
+            if (outstandingRequest != null && !outstandingRequest.isDone()) {
+                log.trace("Cancel outstanding read request for segment {}", asyncInput.getSegmentId());
+                outstandingRequest.cancel(false);
+                outstandingRequest = null;
+            }
+        }
+        if (result != null) {
+            // an event was read successfully; update the cursor position.
+            cursorOffset = readState.offset;
+            log.trace("Read an event @ {}", cursorOffset);
+        }
+    }
+
+    @VisibleForTesting
+    ReadState getReadState() {
+        return readState;
+    }
+
+    static class ReadState {
```

Comment:
This class is not thread safe.

Reply:
The `ReadState` class is internal to the outer class, and all access to it is guarded by `$lock`.

Reply:
This class is not internal to the outer class. It's package-private. This means some other code from this package can call `getReadState` an then mess around with it without holding the enclosing class' lock.

In order to make it thread safe, either make it private or guard all of your members with a `synchronized` block.

Reply:
@andreipaduroiu there's a long history in Java of making private classes be protected only for the purpose of unit testing as is being done here.   `@VisibleForTesting` is a declaration to that effect.  I would not add additional locks, nor would I simplify the unit tests, to accommodate this.

Reply:
If it is visible only for testing, than please annotate it.